internal/pkg/util/exec/starter_linux.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package exec
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/buildcfg"
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines/config"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+// LookStarterPath returns the starter binary path and also returns
+// if the path point to the setuid starter executable
+func LookStarterPath(suid bool, allowedSuid bool) (string, bool) {
+	starterPath := filepath.Join(buildcfg.LIBEXECDIR, "singularity/bin/starter")
+	starterSuid := false
+
+	if buildcfg.SINGULARITY_SUID == 1 && suid {
+		// Singularity was compiled with default option,
+		// checking if setuid binary can be used
+		if !allowedSuid {
+			sylog.Verbosef("Could not use starter suid binary: disabled by configuration")
+			return starterPath, starterSuid
+		}
+		starterSuidPath := fmt.Sprintf("%s-suid", starterPath)
+		if _, err := os.Stat(starterSuidPath); err != nil {
+			// screwed installation, should be pretty rare
+			sylog.Verbosef("Could not use starter suid binary: %s", err)
+			return starterPath, starterSuid
+		}
+		// setuid binary is usable
+		starterPath = starterSuidPath
+		starterSuid = true
+	} else if buildcfg.SINGULARITY_SUID == 0 && suid {
+		// Singularity was compiled with '--without-suid'
+		sylog.Verbosef("Could not use starter suid binary, compiled with '--without-suid' option")
+	}
+
+	return starterPath, starterSuid
+}
+
+// Starter execute starter binary with arguments and pass configuration over pipe
+func Starter(command string, args []string, env []string, config *config.Common) error {
+	pipeEnv, err := SetStarterPipe(config)
+	if err != nil {
+		return err
+	}
+
+	env = append(env, pipeEnv)
+	err = syscall.Exec(command, args, env)
+	if err != nil {
+		return fmt.Errorf("command %s execution failed: %s", command, err)
+	}
+
+	return nil
+}
+
+// StarterCommand creates an exec.Command struct which will execute the starter binary
+func StarterCommand(command string, args []string, env []string, config *config.Common) (*exec.Cmd, error) {
+	pipeEnv, err := SetStarterPipe(config)
+	if err != nil {
+		return nil, err
+	}
+
+	env = append(env, pipeEnv)
+
+	c := &exec.Cmd{
+		Path: command,
+		Args: args,
+		Env:  env,
+	}
+	return c, nil
+}
+
+// setStarterPipe sets a pipe communication channel for JSON configuration data and returns
+// the file descriptor of the read side
+func setStarterPipe(config *config.Common) (int, error) {
+	fd, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM|syscall.SOCK_CLOEXEC, 0)
```

Comment:
since you are touching this... would it be possible to add some comments as to why this is using a socket and not a pipe (given how this function is called "setStarterPipe")?

I'm guessing it's because the communication is bidirectional, but then the question becomes why is it bidirectional?

Any internal documentation regarding that will be appreciated.